examples/text-classification/run_glue.py

```diff
@@ -134,107 +173,98 @@ def main():
     )
 
     # Get datasets
-    train_dataset = (
-        GlueDataset(data_args, tokenizer=tokenizer, cache_dir=model_args.cache_dir) if training_args.do_train else None
-    )
-    eval_dataset = (
-        GlueDataset(data_args, tokenizer=tokenizer, mode="dev", cache_dir=model_args.cache_dir)
-        if training_args.do_eval
-        else None
-    )
-    test_dataset = (
-        GlueDataset(data_args, tokenizer=tokenizer, mode="test", cache_dir=model_args.cache_dir)
-        if training_args.do_predict
-        else None
-    )
-
-    def build_compute_metrics_fn(task_name: str) -> Callable[[EvalPrediction], Dict]:
-        def compute_metrics_fn(p: EvalPrediction):
-            preds = p.predictions[0] if isinstance(p.predictions, tuple) else p.predictions
-            if output_mode == "classification":
-                preds = np.argmax(preds, axis=1)
-            else:  # regression
-                preds = np.squeeze(preds)
-            return glue_compute_metrics(task_name, preds, p.label_ids)
-
-        return compute_metrics_fn
+    datasets = load_dataset("glue", data_args.task_name)
+    sentence1_key, sentence2_key = task_to_keys[data_args.task_name]
+    encode_kwargs = {
+        "tokenizer": tokenizer,
+        "sentence1_key": sentence1_key,
+        "sentence2_key": sentence2_key,
+        "max_length": data_args.max_seq_length,
+    }
+    datasets = datasets.map(preprocess_function, batched=True, fn_kwargs=encode_kwargs)
```

Comment:
I like better to have `preprocess_function` as a closure written just here with the arguments instead of defining it above with kwargs.

It spares the reader a scroll up and down to see what's happening but I understand this is a matter of personal taste.

Reply:
We lose the preprocessing caching for some reason when doing that.